File: basic/01_searching/02_binary_search/BinarySearch.c
/**********************************************************************
 * Simple binary search in C
 * Author : Muhammed Muhsin A. <muhsinibnalazeez@gmail.com>
 *
 * Compile (assuming gcc) : gcc -Wall BinarySearch.c -o BinarySearch.x
 * Run : ./BinarySearch.x
 **********************************************************************/

#include <stdio.h>
#include <stdlib.h>

int binary_search(int [], size_t, int);

int main() {
  int arr[] = {2, 3, 6, 10, 14, 23, 33, 45, 56, 89, 90};
  int key, index;
  size_t size = sizeof(arr)/sizeof(arr[0]);

  key = 33;
  index = binary_search(arr, size, key);
  if(index != -1){
    printf("Key = %d; index = %d\n", key, index);
  }
  else{
    printf("Key = %d; index = NOT FOUND\n", key);
  }
  
  key = 123;
  index = binary_search(arr, size, key);
  if(index != -1){
    printf("Key = %d; index = %d\n", key, index);
  }
  else{
    printf("Key = %d; index = NOT FOUND\n", key);
  }

  return 0;
}

int binary_search(int arr[], size_t n, int key) {
  /*
   * Binary search through sorted array `arr`
   * Assumes sorted array.
  */
  int a, b, c;

  a = 0;
  b = n;
  if(arr[a] == key) return a;
  if(arr[b] == key) return b;
  c = (a + b)/2;
  while(arr[c] != key && c!=a && c!=b) {
    if(arr[c] < key) {
      a = c;
    }
    else if(arr[c] > key) {
      b = c;
    }
    c = (a + b)/2;
  }
  if(arr[c] == key) {
    return c;
  }
  else{
    return -1;
  }


}

File: basic/01_searching/02_binary_search/BinarySearch.java
/**********************************************************************
 * Simple Binary search in Java
 * Author : Muhammed Muhsin A. <muhsinibnalazeez@gmail.com>
 *
 * Compile : javac BinarySearch.java
 * Run : java BinarySearch
 **********************************************************************/

public class BinarySearch {
  static int binarySearch(int[] arr, int key) {
    int a = 0;
    int b = arr.length - 1;
    if(arr[a] == key){
      return a;
    }
    if(arr[b] == key){
      return b;
    }
    int c = (a + b)/2;
    while(arr[c] != key && c > a && c < b){
      if(arr[c] > key){
        b = c;
      }
      else if(arr[c] < key){
        a = c;
      }
      c = (a + b)/2;
    }
    if(arr[c] == key){
      return c;
    }
    return -1;
  }

  public static void main(String[] args){
    int i, key;
    int[] arr = {2, 3, 6, 10, 14, 23, 33, 45, 56, 89, 90};
    key = 33;
    i = binarySearch(arr, key);
    if(i != -1){
      System.out.println("Key = " + key + "; index = " + i);
    }
    else{
      System.out.println("Key = " + key + "; index = NOT FOUND");
    }

    key = 123;
    i = binarySearch(arr, key);
    if(i != -1){
      System.out.println("Key = " + key + "; index = " + i);
    }
    else{
      System.out.println("Key = " + key + "; index = NOT FOUND");
    }
  }
}
